src/main/webapp/app/core/util/alert.service.ts

```diff
@@ -43,104 +119,98 @@ export class AlertService {
+            const translatedMessage = this.translateService.instant(alertInternal.translationKey, alertInternal.translationParams);
             // if translation key exists
-            if (translatedMessage !== `${translationNotFoundMessage}[${alert.translationKey}]`) {
-                alert.message = translatedMessage;
-            } else if (!alert.message) {
-                alert.message = alert.translationKey;
+            if (translatedMessage !== `${translationNotFoundMessage}[${alertInternal.translationKey}]`) {
+                alertInternal.message = translatedMessage;
+            } else if (!alertInternal.message) {
+                alertInternal.message = alertInternal.translationKey;
             }
-        } else if (alert.message) {
+        } else if (alertInternal.message) {
             // Note: in most cases, our code passes the translation key as message
-            alert.message = this.translateService.instant(alert.message, alert.translationParams);
+            alertInternal.message = this.translateService.instant(alertInternal.message, alertInternal.translationParams);
         }
 
-        alert.message = this.sanitizer.sanitize(SecurityContext.HTML, alert.message ?? '') ?? '';
-        alert.timeout = alert.timeout ?? this.timeout;
-        alert.toast = alert.toast ?? this.toast;
-        alert.position = alert.position ?? this.position;
-        alert.close = (alertsArray: Alert[]) => this.closeAlert(alert.id!, alertsArray);
+        alertInternal.message = this.sanitizer.sanitize(SecurityContext.HTML, alertInternal.message ?? '') ?? '';
+        alertInternal.timeout = alertInternal.timeout ?? this.timeout;
+        alertInternal.dismissible = alertInternal.dismissible ?? this.dismissible;
+        alertInternal.close = () => {
+            alertInternal.isOpen = false;
+            const alertIndex = this.alerts.indexOf(alertInternal);
+            if (alertIndex >= 0) {
+                this.alerts.splice(alertIndex, 1);
+                if (alertInternal.onClose) {
+                    alertInternal.onClose(alertInternal);
+                }
+            }
+        };
+        if (alertInternal.action) {
+            alertInternal.action.label = this.sanitizer.sanitize(SecurityContext.HTML, this.translateService.instant(alertInternal.action.label) ?? '') ?? '';
+        }
 
-        (extAlerts ?? this.alerts).push(alert);
+        this.alerts.splice(0, 0, alertInternal);
 
-        if (alert.timeout > 0) {
+        if (alertInternal.timeout > 0) {
             // Workaround protractor waiting for setTimeout.
             // Reference https://www.protractortest.org/#/timeouts
             this.ngZone.runOutsideAngular(() => {
                 setTimeout(() => {
                     this.ngZone.run(() => {
-                        this.closeAlert(alert.id!, extAlerts ?? this.alerts);
+                        alertInternal.close!();
```

Comment:
Why do you need the exclamation mark here?

Reply:
👍